include/CoolPropTools.h

```diff
@@ -341,54 +341,62 @@
         Dictionary(){};
-        bool is_empty(void){return numbers.empty() && strings.empty() && double_vectors.empty() && string_vectors.empty();}
-        void add_string(std::string s1, std::string s2){ strings.insert(std::pair<std::string, std::string>(s1, s2));}
-        void add_number(std::string s1, double d){ numbers.insert(std::pair<std::string, double>(s1, d));}
-        void add_double_vector(std::string s1, std::vector<double> d){ double_vectors.insert(std::pair<std::string, std::vector<double> >(s1, d));}
-        void add_string_vector(std::string s1, std::vector<std::string> d){ string_vectors.insert(std::pair<std::string, std::vector<std::string> >(s1, d));}
-        std::string get_string(std::string s)
+        bool is_empty(void) const {return numbers.empty() && strings.empty() && double_vectors.empty() && string_vectors.empty();}
+        void add_string(const std::string &s1, const std::string &s2){ strings.insert(std::pair<std::string, std::string>(s1, s2));}
+        void add_number(const std::string &s1, double d){ numbers.insert(std::pair<std::string, double>(s1, d));}
+        void add_double_vector(const std::string &s1, const std::vector<double> &d){ double_vectors.insert(std::pair<std::string, std::vector<double> >(s1, d));}
+        void add_string_vector(const std::string &s1, const std::vector<std::string> &d){ string_vectors.insert(std::pair<std::string, std::vector<std::string> >(s1, d));}
+        std::string get_string(const std::string &s) const
         {
-            if (strings.find(s) != strings.end()){ 
-                return strings[s]; 
-            } 
-            else{ 
-                throw CoolProp::ValueError(format("%s could not be matched in get_string",s.c_str())); 
+            strings_map::const_iterator i = strings.find(s);
+            if (i != strings.end()){
+                return i->second;
+            }
+            else{
+                throw CoolProp::ValueError(format("%s could not be matched in get_string",s.c_str()));
             }
         };
-        double get_double(std::string s)
+        double get_double(const std::string &s) const
         {
-            if (numbers.find(s) != numbers.end()){ 
-                return numbers[s]; 
-            } 
-            else{ 
-                throw CoolProp::ValueError(format("%s could not be matched in get_number",s.c_str())); 
+            numbers_map::const_iterator i = numbers.find(s);
+            if (i != numbers.end()){
+                return i->second;
+            }
+            else{
+                throw CoolProp::ValueError(format("%s could not be matched in get_number",s.c_str()));
             }
         };
-        double get_number(std::string s)
+        double get_number(const std::string &s) const
         {
             return get_double(s);
         };
-        const std::vector<double>& get_double_vector(std::string s)
+        const std::vector<double>& get_double_vector(const std::string &s) const
```

Comment:
yes, and it was three-fold here.
1. I noticed the param-by-value Dictionary (ReducingFunctions.h line 327). The dictionary has four maps, that contain numbers, strings, and vectors of numbers and strings. Each has to be recursively copied, and released after use, and each object has its constructors and destructors. So, naturally, I wanted to pass it by const ref. But when I did that, the error appeared that the object cannot be **const**, as the methods called are not const.
2. I see that methods should not make changes to Dictionary - they are simple selectors. So I go and change their declarations as const, too. But compiler complains that as the method is const, it's illegal to call non-const methods of members there. I see, that the methods are implemented this way: first, the key is searched using **find()**, and only if found, the **operator []** is called, that makes search again. This is not optimal per se, but additionally, in maps, **operator []** is not const: it looks for the key, and if not found, it automatically adds new entry for this key, thus changing the object. This helped me to notice this inefficient implementation. So, I simplified things by making only one lookup using const method **find()**.
3. Further, it **may** help compiler generate more efficient code internally, even in places where the code is unchanged, as now it has a hint that the object's state stays unchanged.